syntax = "proto3";

package humanoid_robot.PB.interfaces;

// Import the base variant types
import "common/variant.proto";

// Common Request execute information
message ResultStatus {
  string code = 1;        // 错误代码
  string message = 2;     // 错误消息
  string details = 3;     // 详细信息
  int64 timestamp = 4;    // 错误发生时间戳
}

// Common parameter structure using Dictionary
message Params {
  humanoid_robot.PB.common.Dictionary keyValueList = 1;  // 通用参数字典
}

// =============================================================================
// Unified Operations (gRPC supports both sync/async automatically)
// =============================================================================

// Send Request
message SendRequest {
  humanoid_robot.PB.common.Dictionary input = 1;  // 创建请求数据
  humanoid_robot.PB.common.Dictionary params = 2;                      // 附加参数
}

// Send Response
message SendResponse {
  humanoid_robot.PB.common.Dictionary output = 1; // Action操作信息
  ResultStatus ret = 2;                    // 返回码
}

// Subscribe Request
message SubscribeRequest {
  humanoid_robot.PB.common.Dictionary input = 1;  // 创建请求数据
  humanoid_robot.PB.common.Dictionary params = 2;                      // 附加参数
}

// Subscribe Response
message SubscribeResponse {
  humanoid_robot.PB.common.Dictionary output = 1; // Action操作信息
  ResultStatus ret = 2;                    // 返回码
}

// Unsubscribe Request
message UnsubscribeRequest {
  humanoid_robot.PB.common.Dictionary input = 1;  // 创建请求数据
  humanoid_robot.PB.common.Dictionary params = 2;                      // 附加参数
}

// Unsubscribe Response
message UnsubscribeResponse {
  humanoid_robot.PB.common.Dictionary output = 1; // Action操作信息
  ResultStatus ret = 2;                    // 返回码
}

// Action Request
message ActionRequest {
  humanoid_robot.PB.common.Dictionary input = 1; // Action操作信息
  humanoid_robot.PB.common.Dictionary params = 2;                      // 附加参数
}

// Action Response
message ActionResponse {
  humanoid_robot.PB.common.Dictionary output = 1; // Action操作信息
  ResultStatus ret = 2;                    // 返回码
}

// Query Request
message QueryRequest {
  humanoid_robot.PB.common.Dictionary input = 1;  // 创建请求数据
  humanoid_robot.PB.common.Dictionary params = 2;                      // 附加参数
}

// Query Response
message QueryResponse {
  humanoid_robot.PB.common.Dictionary output = 1; // Action操作信息
  ResultStatus ret = 2;                    // 返回码
}
